plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Definition.hs

```diff
@@ -319,6 +323,236 @@ test_IdBuiltinData =
+    evals False GreaterThanByteString [cons @ByteString "" , cons @ByteString "mpla"]
+
+    -- strings
+    evals @String "a" CharToString [cons 'a']
+    evals @String "mpla" Append [cons @String "", cons @String "mpla"]
+    evals False EqualsString [cons @String "" , cons @String "mpla"]
+    evals True EqualsString [cons @String "mpla" , cons @String "mpla"]
+    evals @String "hello world" Append [cons @String "hello", cons @String " world"]
+
+    -- id for subset char8 of utf8
+    evals @ByteString "hello world" EncodeUtf8 [cons @String "hello world"]
+    evals @String "hello world" DecodeUtf8 [cons @ByteString "hello world"]
+
+    -- the 'o's replaced with greek o's, so they are kind of "invisible"
+    evals @ByteString "hell\191 w\191rld" EncodeUtf8 [cons @String "hellο wοrld"]
+    -- cannot decode back, because bytestring only works on Char8 subset of utf8
+    evals @String "hell\191 w\191rld" DecodeUtf8 [cons @ByteString "hell\191 w\191rld"]
+
+
+-- | Test all list-related builtins
+test_List :: TestTree
+test_List = testCase "List" $ do
+    evalsL False NullList integer [cons @[Integer] [1,2]]
+    evalsL False NullList integer [cons @[Integer] [1]]
+    evalsL True NullList integer [cons @[Integer] []]
+
+    evalsL @Integer 1 HeadList integer [cons @[Integer] [1,3]]
+    evalsL @[Integer] [3,4,5] TailList integer [cons @[Integer] [1,3,4,5]]
+
+    failsL HeadList integer [cons @[Integer] []]
+    failsL TailList integer [cons @[Integer] []]
+
+    evalsL @[Integer] [1] MkCons integer [cons @Integer 1, cons @[Integer] []]
+    evalsL @[Integer] [1,2] MkCons integer [cons @Integer 1, cons @[Integer] [2]]
+
+    -- ChooseList with zero scrutinee
+    let chooseListZ = mkIterApp ()
+                      (tyInst () (apply () (tyInst () Builtin.caseList integer) $ cons @[Integer] []) bool)
+                      [ -- zero
+                        true
+                        -- cons
+                      , runQuote $ do
+                              a1 <- freshName "a1"
+                              a2 <- freshName "a2"
+                              pure $ lamAbs () a1 integer $ lamAbs () a2 (TyApp () Builtin.list integer) false
+                      ]
+    Right (EvaluationSuccess true)  @=?  typecheckEvaluateCekNoEmit defaultCekParameters chooseListZ
+
+    -- ChooseList with cons scrutinee
+    let chooseListC = mkIterApp ()
+                      (tyInst () (apply () (tyInst () Builtin.caseList integer) $ cons @[Integer] [1,2]) bool)
+                      [ -- zero
+                        false
+                        -- cons
+                      , runQuote $ do
+                              a1 <- freshName "a1"
+                              a2 <- freshName "a2"
+                              pure $ lamAbs () a1 integer $ lamAbs () a2 (TyApp () Builtin.list integer) true
+                      ]
+    Right (EvaluationSuccess true)  @=?  typecheckEvaluateCekNoEmit defaultCekParameters chooseListC
```

Comment:
I'd abstract this pattern given that you use it twice. And I wouldn't switch the order of `false` and `true` and instead make it visible that the results changes, but that's nitpicking.